File: javaQuarkusShip/src/main/java/pl/battleships/javaquarkusship/api/GameResource.java
package pl.battleships.javaquarkusship.api;


import io.quarkus.logging.Log;
import org.jboss.resteasy.reactive.RestResponse;
import org.jboss.resteasy.reactive.server.ServerExceptionMapper;
import pl.battleships.javaquarkusship.api.dto.GameDto;
import pl.battleships.javaquarkusship.api.dto.GameStatusDto;
import pl.battleships.javaquarkusship.api.dto.PositionDto;
import pl.battleships.javaquarkusship.api.dto.ShotStatusDto;
import pl.battleships.core.exception.*;
import pl.battleships.javaquarkusship.service.GameService;


import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.ws.rs.core.Response;
import java.util.List;

@ApplicationScoped
public class GameResource implements GameApi {

    @Inject
    GameService gameService;

    @Override
    public List<PositionDto> getAllShots(String id) {
        return gameService.getAllShots(id);
    }

    @Override
    public GameStatusDto getGameStatus(String id) {
        return gameService.getGameStatus(id);
    }

    @Override
    public void joinGame(GameDto gameDto) {
        gameService.joinTheGame(gameDto);
    }

    @Override
    public ShotStatusDto shot(String id, PositionDto positionDto) {
        return gameService.opponentShot(id, positionDto);
    }

    @ServerExceptionMapper
    public RestResponse<Void> duplicatedGameException(DuplicatedGameException x) {
        return RestResponse.status(Response.Status.CONFLICT);
    }

    @ServerExceptionMapper
    public RestResponse<Void> noGameFoundException(NoGameFoundException x) {
        return RestResponse.status(Response.Status.NOT_FOUND);
    }

    @ServerExceptionMapper
    public RestResponse<Void> gameOver(GameOverException x) {
        Log.info("Game over");
        return RestResponse.status(Response.Status.GONE);
    }

    @ServerExceptionMapper
    public RestResponse<Void> invalidParams(InvalidParamException x) {
        return RestResponse.status(422);
    }

    @ServerExceptionMapper
    public RestResponse<Void> invalidMoveException(InvalidMoveException x) {
        return RestResponse.status(Response.Status.FORBIDDEN);
    }

}

File: kotlinQuarkusShip/src/main/kotlin/pl/battleships/kotlinquarkusship/resource/GameResource.kt
package pl.battleships.kotlinquarkusship.resource

import org.jboss.resteasy.reactive.server.ServerExceptionMapper
import pl.battleships.api.GameApi
import pl.battleships.api.GameStatusApi
import pl.battleships.api.dto.GameDto
import pl.battleships.api.dto.GameStatusDto
import pl.battleships.api.dto.PositionDto
import pl.battleships.api.dto.ShotStatusDto
import pl.battleships.core.exception.*
import pl.battleships.kotlinquarkusship.service.GameService
import javax.enterprise.context.ApplicationScoped
import javax.ws.rs.core.Response

@ApplicationScoped
class GameResource(
    private val gameService: GameService
) : GameApi, GameStatusApi {
    override fun getAllShots(gameId: String): List<PositionDto> {
        return gameService.getAllShots(gameId)
    }

    override fun joinGame(game: GameDto) {
        gameService.joinTheGame(game)
    }

    override fun shot(gameId: String, position: PositionDto): ShotStatusDto {
        return gameService.opponentShot(gameId, position)
    }

    @ServerExceptionMapper
    fun duplicatedGameException(x: DuplicatedGameException?): Response {
        return Response.status(Response.Status.CONFLICT).build()
    }

    @ServerExceptionMapper
    fun noGameFoundException(x: NoGameFoundException?): Response {
        return Response.status(Response.Status.NOT_FOUND).build()
    }

    @ServerExceptionMapper
    fun gameOver(x: GameOverException?): Response {
        return Response.status(Response.Status.GONE).build()
    }

    @ServerExceptionMapper
    fun invalidParams(x: InvalidParamException?): Response {
        return Response.status(422).build()
    }

    @ServerExceptionMapper
    fun invalidMoveException(x: InvalidMoveException?): Response {
        return Response.status(Response.Status.FORBIDDEN).build()
    }

    override fun getGameStatus(id: String): GameStatusDto {
        return gameService.getGameStatus(id)
    }
}